src/main/java/games/strategy/debug/LogReader.java

```diff
@@ -0,0 +1,58 @@
+package games.strategy.debug;
+
+import java.io.ByteArrayOutputStream;
+import java.io.PrintStream;
+import java.util.function.BooleanSupplier;
+
+import javax.swing.JTextArea;
+
+final class LogReader {
+  private final JTextArea text;
+  private final ListeningByteArrayOutputStream stream = new ListeningByteArrayOutputStream();
+  private final BooleanSupplier displayConsoleOnWriteSupplier;
+  private final GenericConsole parentConsole;
+  private final PrintStream out;
+
+  LogReader(
+      final PrintStream out,
+      final JTextArea text,
+      final BooleanSupplier displayConsoleOnWriteSupplier,
+      final GenericConsole parentConsole) {
+    this.out = out;
+    this.text = text;
+    this.displayConsoleOnWriteSupplier = displayConsoleOnWriteSupplier;
+    this.parentConsole = parentConsole;
+  }
+
+  /**
+   * Returns an OutputStream that triggers the Console (when enabled) when something writes to it.
+   */
+  ByteArrayOutputStream getStream() {
+    return stream;
+  }
+
+  private void addConsoleText() {
+    text.append(new String(stream.toByteArray()));
+    if (displayConsoleOnWriteSupplier.getAsBoolean()) {
+      parentConsole.setVisible(true);
```

Comment:
We have two UI methods being called here.  Is there a guarantee this method is always invoked on the EDT?

Kind of related, but something's niggling my brain about this method being called while the `ListeningByteArrayOutputStream` lock is held.  Maybe a way to address both issues is to have this method extract the text from the stream while the lock is held, and then run a closure on the EDT that actually manipulates the UI controls.  Something like:

```java
  private void addConsoleText() {
    final String consoleText = stream.toString();
    stream.reset();
    SwingUtilities.invokeLater(() -> {
      text.append(consoleText);
      if (displayConsoleOnWriteSupplier.getAsBoolean()) {
        parentConsole.setVisible(true);
      }
    });
  }
```

But that might break the unit tests on Travis because it will require an EDT to be spun up.  I could be wrong about that, though.